sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1024,6 +1024,27 @@ class DatasetSuite extends QueryTest
     assert(namedObservation.get === expected)
   }
 
+  test("SPARK-45656: named observations with the same name on different datasets") {
+    val namedObservation1 = Observation("named")
+    val df1 = spark.range(50)
+    val observed_df1 = df1.observe(
+      namedObservation1, count(lit(1)).as("count"))
```

Comment:
does the `observe` with string input have the same problem? or it's only an issue with `Observation`?

Reply:
If `name` in `Dataset.observe(name, expr, exprs)` is not unique across the spark session, then yes, we have the same problem. `Observation(name)` inherits the problem from `Dataset.observe(name, expr, exprs)`. `Observation()` avoids this problem.

Reply:
There is a rule that enforces that `name` is unique within a plan (except for identical expressions: self-joins).

Could we have some similar check across the spark session?